### PR TITLE
Make `MesonNinja` respect the toolchainopts with buildtype as well as `--debug` and `--optimization` flags

### LINKED ISSUE #3280: set a default buildtype in MesonNinja easyblock
for e.g. CMake we add `-DCMAKE_BUILD_TYPE=Release` by default. 
We don't currently do anything like that for meson: https://github.com/easybuilders/easybuild-easyblocks/blob/develop/easybuild/easyblocks/generic/mesonninja.py#L95

this means we actually default to `debug` without optimisations ([ref](https://mesonbuild.com/Builtin-options.html#core-options)):
|Option|Default value|Description|Is per machine|Is per subproject|
|---|---|---|---|---|
|buildtype {plain, debug, debugoptimized, release, minsize, custom}|debug|Build type to use|no|no|

these options correspond to ([ref](https://mesonbuild.com/Builtin-options.html#details-for-buildtype)):
|buildtype|debug|optimization|
|---|---|---|
|plain|false|plain|
|debug|true|0
|debugoptimized|true|2|
|release|false|3|
|minsize|true|s|

### PR DESCRIPTION
**Pull request by Micket**
Fixes #3280 

Should also merge https://github.com/easybuilders/easybuild-easyconfigs/pull/21619 once this is merged.

In this version, i double up on both buildtype as well as --optimizationand --debug flags to respect the toolchainopts. I think this will be the better approach, but i haven't tested anything yet.

### CODE REVIEW

**Review by boegel — DISMISSED** [review]

**Comment by Micket** [comment]
> WARNING: Recommend using either -Dbuildtype or -Doptimization + -Ddebug. Using both is redundant since they override each other. See: https://mesonbuild.com/Builtin-options.html#build-type-options

ok, so apparently meson doesn't want you to do this. I find this strange, as just using the optimization + debug options wouldn't cover things like -DNODEBUG defines. Maybe noone using meson uses such defines in their C/C++ codes? :/

**Comment by Micket** [comment]
https://github.com/mesonbuild/meson-python/pull/325 No this seems to indicate otherwise. We certainly always want the NDEBUG set for our builds...

**Comment by Micket** [comment]
Alright, so to summarize

1. I add the -v flag so we can see what we are compiling
2. I set the build type, in case they have some "if-release" logic that adds extra flags and such.
3. I set optimization and debug on top of that, which produces a warning that I disagree with due to the point mentioned above.
4. I set the `-bn_debug=true` flag for all builds that aren't noopt, so that we get -DNDEBUG when building debugoptimized as well as release. I think by default, the meson-python actually adds the flag `-bn_debug=if-release` but if we are enabling debug symbols, you don't want to squash this. They mention in the issue tracking n_debug flag that they don't add it automatically to release because they expect packagers to set all their flags themselves anyway.. and i guess they are right (now).

All in all this brings meson builds in line with the rest of the stuff we build, and I don't think any of it is controversial. 
Wouldn't surprise me if we see a large speedup in some packages after these changes, but it's hard to test things like "cairo" or "Wayland" for performance.

**Comment by Micket** [comment]
Test report by @Micket
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _cairo-1.18.0-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _GLib-2.78.1-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _libglvnd-1.7.0-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _Wayland-1.23.0-GCCcore-13.3.0.eb_ 

Build succeeded for 4 out of 4 (4 easyconfigs in total)
vera-skylake-build - Linux Rocky Linux 8.9, x86_64, Intel Xeon Processor (Skylake, IBRS, no TSX), Python 3.6.8
See https://gist.github.com/Micket/1c22578d1e11a2d81ddd7f38a3c0030e for a full test report.

**Comment by Micket** [comment]
Test report by @Micket
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _cairo-1.18.0-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _GLib-2.78.1-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _libglvnd-1.7.0-GCCcore-13.3.0.eb_ 
 * **FAIL (build issue)** _Wayland-1.23.0-GCCcore-13.3.0.eb_ (partial log available at https://gist.github.com/Micket/b25cd2d50bfff9f3de75fde00bbb16a9)

Build succeeded for 3 out of 4 (4 easyconfigs in total)
vera-skylake-build - Linux Rocky Linux 8.9, x86_64, Intel Xeon Processor (Skylake, IBRS, no TSX), Python 3.6.8
See https://gist.github.com/Micket/d182683aa7be4718a6ad9167888a51d9 for a full test report.

**Comment by Micket** [comment]
So, Wayland apparently needs to be built with asserts, else it fails the test suite. Which means, one can't actually test  a true install... Not sure how much it impacts performance, or if we care about tools like wayland for that sort of performance in HPC settings, so I'm making n_debug into an option.

**Comment by Micket** [comment]
OK, so meson is really strange when specifying both buildtype and optmization/debug flags. 
They claim that they mutually exclusive, but I disagree; the value of "buildtype" itself has value given things like
```
if get_option('buildtype') == 'release':
   # custom logic here
```

In particular `-Db_ndebug=if-release`  is something a lot of things set (all of meson-python projects do this by default), so, it's not some little thing either. Though, if-release also leaves a lot to be desired; what about debugoptimized? is that supposed to be a release-like thing? Again `debug` is being used vaguely here.

I checked the meson.build from all the stuff that uses MesonNinja to see what, if any, such custom logic anyone might be doing, and apart from some harmless flags and b_ndebug=if-release, I couldn't find anything that worried me, so I'm going to go with the option of making buildtype and optimization/debug mutually exclusive for our easyblock as well, and just rely on optimization, debug, and b_ndebug flags.

**Comment by Micket** [comment]
Test report by @Micket
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ATK-2.38.0-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _at-spi2-atk-2.38.0-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _at-spi2-core-2.50.0-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _cairo-1.18.0-GCCcore-13.3.0.eb_ 
 * **FAIL (build issue)** _cairomm-1.16.2-GCC-12.3.0.eb_ (partial log available at https://gist.github.com/Micket/21e0eee0053815efb7ebedf48fbbfef8)
 * **SUCCESS** _desktop-file-utils-0.27-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _DFT-D4-3.6.0-intel-2022a.eb_ 
 * **SUCCESS** _evince-45.0-GCC-12.3.0.eb_ 
 * **SUCCESS** _freebayes-1.3.7-gfbf-2023a-R-4.3.2.eb_ 
 * **SUCCESS** _FUSE-3.16.2-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _Gdk-Pixbuf-2.42.10-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _GLib-2.80.4-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _GLibmm-2.78.1-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _glib-networking-2.72.1-GCCcore-11.3.0.eb_ 
 * **SUCCESS** _GObject-Introspection-1.78.1-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _Graphene-1.10.8-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _GST-plugins-bad-1.22.5-GCC-12.3.0.eb_ 
 * **SUCCESS** _GST-plugins-base-1.22.5-GCC-12.3.0.eb_ 
 * **SUCCESS** _GStreamer-1.22.5-GCC-12.3.0.eb_ 
 * **SUCCESS** _GTK3-3.24.39-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _GTK4-4.13.1-GCC-12.3.0.eb_ 
 * **SUCCESS** _inih-57-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _libdrm-2.4.122-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _libepoxy-1.5.10-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _libGLU-9.0.3-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _libglvnd-1.7.0-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _libgxps-0.3.2-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _libhandy-1.8.2-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _libpciaccess-0.18.1-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _LibSoup-3.0.8-GCC-11.3.0.eb_ 
 * **SUCCESS** _mm-common-1.0.6-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _MODFLOW-6.4.4-foss-2023a.eb_ 
 * **SUCCESS** _Pango-1.51.0-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _pixman-0.43.4-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _PretextMap-0.1.9-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _PyCairo-1.25.1-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _Wayland-1.23.0-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _wpebackend-fdo-1.14.1-GCCcore-11.3.0.eb_ 
 * **SUCCESS** _X11-20240607-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _xtb-6.7.0-gfbf-2023a.eb_ 
 * **SUCCESS** _Xvfb-21.1.9-GCCcore-13.2.0.eb_ 

Build succeeded for 40 out of 41 (41 easyconfigs in total)
vera-skylake-build - Linux Rocky Linux 8.9, x86_64, Intel Xeon Processor (Skylake, IBRS, no TSX), Python 3.6.8
See https://gist.github.com/Micket/49102596d81d8d925377b46e82b7475f for a full test report.

**Comment by Micket** [comment]
The error on cairomm is something unrelated, 2 tests crashes for me no matter what easyblock i use, even in 4.9.2, so it shouldn't let this block the merge.

**Review by jfgrimm — APPROVED** [review]
lgtm

File: easybuild/easyblocks/generic/mesonninja.py [hunk 1]
```diff
@@ -53,12 +53,26 @@ def extra_options(extra_vars=None):
...
+    @property
+    def optimization(self):
+        """Optimization level"""
+        if self.toolchain.options.get('noopt', False):
+            return 0
+        elif self.toolchain.options.get('lowopt', False):
+            return 1
+        else:
+            return 2
+
```
Comment (jfgrimm) [inline easybuild/easyblocks/generic/mesonninja.py:75]: wait, what if we have `'opt': True` (as in, use `-O3`)?
surely this needs another elif?

**Review by jfgrimm — CHANGES_REQUESTED** [review]

File: easybuild/easyblocks/generic/mesonninja.py (hunk 1 above)
Comment (jfgrimm) [reply easybuild/easyblocks/generic/mesonninja.py:75]:
```suggestion
    @property
    def optimization(self):
        """Optimization level"""
        if self.toolchain.options.get('noopt', False):
            return 0
        elif self.toolchain.options.get('lowopt', False):
            return 1
        elif self.toolchain.options.get('opt', False):
            return 3
        else:
            return 2

```

**Review by jfgrimm — APPROVED** [review]

**Pull request merged**